From f6345ce2ce4c4795dcff8c33c198d07145e8a52d Mon Sep 17 00:00:00 2001
From: Ming Qian <ming.qian@nxp.com>
Date: Fri, 10 Jun 2022 17:00:02 +0800
Subject: [PATCH 60/91] MMFMWK-9186-4 video: Add support for BGRA_12LE

BGRA_12LE is reverse RGB with alpha channel last, 12 bits per component,
expanded to 16bits.
Data in the 12 high bits, zeros in the 4 low bits,
arranged in little endian order.

Upstream-Status: Pending

Conflicts:
	gst-libs/gst/video/video-format.h
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 gst-libs/gst/video/video-converter.c |  1 +
 gst-libs/gst/video/video-format.c    | 58 ++++++++++++++++++++++++++++
 gst-libs/gst/video/video-format.h    | 15 ++++++-
 gst-libs/gst/video/video-info.c      |  5 +++
 4 files changed, 77 insertions(+), 2 deletions(-)

diff --git a/gst-libs/gst/video/video-converter.c b/gst-libs/gst/video/video-converter.c
index 20d4c01..f51bb58 100644
--- a/gst-libs/gst/video/video-converter.c
+++ b/gst-libs/gst/video/video-converter.c
@@ -7330,6 +7330,7 @@ get_scale_format (GstVideoFormat format, gint plane)
     case GST_VIDEO_FORMAT_GBR_16BE:
     case GST_VIDEO_FORMAT_GBR_16LE:
     case GST_VIDEO_FORMAT_BGR_12LE:
+    case GST_VIDEO_FORMAT_BGRA_12LE:
     case GST_VIDEO_FORMAT_NV12_64Z32:
     case GST_VIDEO_FORMAT_NV12_4L4:
     case GST_VIDEO_FORMAT_NV12_32L32:
diff --git a/gst-libs/gst/video/video-format.c b/gst-libs/gst/video/video-format.c
index dd1a7d2..2b71814 100644
--- a/gst-libs/gst/video/video-format.c
+++ b/gst-libs/gst/video/video-format.c
@@ -3538,6 +3538,62 @@ pack_BGR_12LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
   }
 }
 
+#define PACK_BGRA_12LE GST_VIDEO_FORMAT_ARGB64, unpack_BGRA_12LE, 1, pack_BGRA_12LE
+static void
+unpack_BGRA_12LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    gpointer dest, const gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], gint x, gint y, gint width)
+{
+  int i;
+  const guint16 *restrict s = GET_LINE (y);
+  guint16 *d = dest, B, G, R, A;
+
+  s += x * 4;
+
+  for (i = 0; i < width; i++) {
+    B = GST_READ_UINT16_LE (s + 4 * i + 0) & 0xfff0;
+    G = GST_READ_UINT16_LE (s + 4 * i + 1) & 0xfff0;
+    R = GST_READ_UINT16_LE (s + 4 * i + 2) & 0xfff0;
+    A = GST_READ_UINT16_LE (s + 4 * i + 3) & 0xfff0;
+
+    if (!(flags & GST_VIDEO_PACK_FLAG_TRUNCATE_RANGE)) {
+      R |= (R >> 12);
+      G |= (G >> 12);
+      B |= (B >> 12);
+      A |= (A >> 12);
+    }
+
+    d[i * 4 + 0] = A;
+    d[i * 4 + 1] = R;
+    d[i * 4 + 2] = G;
+    d[i * 4 + 3] = B;
+  }
+}
+
+static void
+pack_BGRA_12LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
+    const gpointer src, gint sstride, gpointer data[GST_VIDEO_MAX_PLANES],
+    const gint stride[GST_VIDEO_MAX_PLANES], GstVideoChromaSite chroma_site,
+    gint y, gint width)
+{
+  int i;
+  guint16 *restrict d = GET_LINE (y);
+  const guint16 *restrict s = src;
+  guint16 B, G, R, A;
+
+  for (i = 0; i < width; i++) {
+    A = s[i * 4 + 0];
+    R = s[i * 4 + 1];
+    G = s[i * 4 + 2];
+    B = s[i * 4 + 3];
+
+    GST_WRITE_UINT16_LE (d + 4 * i + 0, B);
+    GST_WRITE_UINT16_LE (d + 4 * i + 1, G);
+    GST_WRITE_UINT16_LE (d + 4 * i + 2, R);
+    GST_WRITE_UINT16_LE (d + 4 * i + 3, A);
+  }
+}
+
 #define PACK_Y444_10LE GST_VIDEO_FORMAT_AYUV64, unpack_Y444_10LE, 1, pack_Y444_10LE
 static void
 unpack_Y444_10LE (const GstVideoFormatInfo * info, GstVideoPackFlags flags,
@@ -7902,6 +7958,8 @@ static const VideoFormat formats[] = {
       PSTR222, PLANE0, OFFS0, SUB444, PACK_Y312_LE),
   MAKE_RGB_LE_FORMAT (BGR_12LE, "raw video", DPTH12_12_12_HI, PSTR222, PLANE0,
       OFFS0, SUB444, PACK_BGR_12LE),
+  MAKE_RGB_LE_FORMAT (BGRA_12LE, "raw video", DPTH12_12_12_12_HI, PSTR8888, PLANE0,
+      OFFS0, SUB444, PACK_BGRA_12LE),
 };
 
 G_GNUC_END_IGNORE_DEPRECATIONS;
diff --git a/gst-libs/gst/video/video-format.h b/gst-libs/gst/video/video-format.h
index 561f087..09512ac 100644
--- a/gst-libs/gst/video/video-format.h
+++ b/gst-libs/gst/video/video-format.h
@@ -172,6 +172,7 @@ G_BEGIN_DECLS
  * @GST_VIDEO_FORMAT_Y012_LE: 12-bit grayscale, least significant byte first, expanded to 16bits, zeros in the 4 low bits,  (Since: 1.24)
  * @GST_VIDEO_FORMAT_Y312_LE: packed 4:4:4 YUV, 12 bits per channel (Y-U-V) (Since: 1.24)
  * @GST_VIDEO_FORMAT_BGR_12LE: reverse RGB, 12 bits per channel (Since: 1.24)
+ * @GST_VIDEO_FORMAT_BGRA_12LE: reverse RGB with alpha channel last, 12 bits per channel (Since: 1.24)
  *
  * Enum value describing the most common video formats.
  *
@@ -652,6 +653,16 @@ typedef enum {
    * Since: 1.24
    */
   GST_VIDEO_FORMAT_BGR_12LE,
+
+  /**
+   * GST_VIDEO_FORMAT_BGRA_12LE:
+   *
+   * Reverse RGB with alpha channel last, 12 bits little endian.
+   * expanded to 16 bits, data in the 12 high bits, zeros in the 4 low bits
+   *
+   * Since: 1.24
+   */
+  GST_VIDEO_FORMAT_BGRA_12LE,
 } GstVideoFormat;
 
 #define GST_VIDEO_MAX_PLANES 4
@@ -1136,7 +1147,7 @@ gconstpointer  gst_video_format_get_palette          (GstVideoFormat format, gsi
     "Y444, BGRP, GBR, RGBP, NV24, v308, IYU2, RGBx, xRGB, BGRx, xBGR, RGB, " \
     "BGR, Y42B, NV16, NV61, YUY2, YVYU, UYVY, VYUY, I420, YV12, NV12, NV21, " \
     "NV12_16L32S, NV12_32L32, NV12_4L4, NV12_64Z32, NV12_8L128, Y41B, IYU1, " \
-    "YUV9, YVU9, BGR16, RGB16, BGR15, RGB15, RGB8P, BGR_12LE, GRAY16_BE, GRAY16_LE, " \
+    "YUV9, YVU9, BGR16, RGB16, BGR15, RGB15, RGB8P, BGR_12LE, BGRA_12LE, GRAY16_BE, GRAY16_LE, " \
     "GRAY10_LE32, GRAY8"
 #elif G_BYTE_ORDER == G_LITTLE_ENDIAN
 #define GST_VIDEO_FORMATS_ALL_STR "A444_16LE, A444_16BE, AYUV64, RGBA64_LE, " \
@@ -1155,7 +1166,7 @@ gconstpointer  gst_video_format_get_palette          (GstVideoFormat format, gsi
     "BGRP, GBR, RGBP, NV24, v308, IYU2, RGBx, xRGB, BGRx, xBGR, RGB, BGR, " \
     "Y42B, NV16, NV61, YUY2, YVYU, UYVY, VYUY, I420, YV12, NV12, NV21, " \
     "NV12_16L32S, NV12_32L32, NV12_4L4, NV12_64Z32, NV12_8L128, Y41B, IYU1, " \
-    "YUV9, YVU9, BGR16, RGB16, BGR15, RGB15, RGB8P, BGR_12LE, GRAY16_LE, GRAY16_BE, " \
+    "YUV9, YVU9, BGR16, RGB16, BGR15, RGB15, RGB8P, BGR_12LE, BGRA_12LE, GRAY16_LE, GRAY16_BE, " \
     "GRAY10_LE32, GRAY8"
 #endif
 
diff --git a/gst-libs/gst/video/video-info.c b/gst-libs/gst/video/video-info.c
index 1b23f04..42a74e5 100644
--- a/gst-libs/gst/video/video-info.c
+++ b/gst-libs/gst/video/video-info.c
@@ -869,6 +869,11 @@ fill_planes (GstVideoInfo * info, gsize plane_size[GST_VIDEO_MAX_PLANES])
       info->offset[0] = 0;
       info->size = info->stride[0] * height;
       break;
+    case GST_VIDEO_FORMAT_BGRA_12LE:
+      info->stride[0] = width * 8;
+      info->offset[0] = 0;
+      info->size = info->stride[0] * height;
+      break;
     case GST_VIDEO_FORMAT_GRAY8:
       info->stride[0] = GST_ROUND_UP_4 (width);
       info->offset[0] = 0;
-- 
2.43.0

